src/services/data/data.service.ts

```diff
@@ -7,40 +6,75 @@ import {
   revenueMeasure,
   sourceCategory,
 } from '../../models/data-cube/presets';
+import { DataCube } from '../../models/data-cube/data-cube.model';
 import { betweenDates } from '../../models/data-cube/filters';
 import { generateCube } from 'src/models/data-cube/generation';
-import { Datum } from '../../d3/xy-chart.d3';
+import { OnDestroy } from '@angular/core';
+import { map, takeUntil, throttleTime } from 'rxjs/operators';
+import { asyncScheduler, combineLatest, Observable, ReplaySubject, Subject } from 'rxjs';
+import { PreferenceService } from '../preference/preference.service';
+import { TimeSeriesQueryOptions } from './types';
+import { QueryOptions } from '../../models/data-cube/types';
 
-export class DataService {
-  private dataCube = generateCube(
-    [countryCategory, browserCategory, sourceCategory],
-    [activeUserMeasure, revenueMeasure, eventCountMeasure],
-    {
+export class DataService implements OnDestroy {
+  private dataCube$ = new ReplaySubject<DataCube>(1);
+  private destroy$ = new Subject();
+
+  constructor(private preferenceService: PreferenceService) {
+    const categories = [countryCategory, browserCategory, sourceCategory];
+    const measures = [activeUserMeasure, revenueMeasure, eventCountMeasure];
+    const defaultSettings = {
       avgHits: 10000,
       hitStdDev: 100,
       avgUsers: 100,
       userStdDev: 1,
       avgSessionsPerUser: 5,
       sessionsPerUserStdDev: 3,
-    },
-  );
+    };
+
+    this.preferenceService.data$
+      .pipe(takeUntil(this.destroy$))
+      .pipe(throttleTime(500, asyncScheduler, { leading: true, trailing: true }))
+      .subscribe(preference => {
+        this.dataCube$.next(generateCube(
+          categories,
+          measures,
+          {
+            ...defaultSettings,
+            ...preference,
+          }));
+      });
+  }
 
-  getMeasureOverDays(measureName: string, days = 30): Datum[] {
-    const categoryName = 'nthDay';
-    const endDate = DateTime.local();
-    const startDate = endDate.minus({ day: days });
+  ngOnDestroy() {
+    this.destroy$.next();
+    this.destroy$.complete();
+  }
+
+  observeData(queryOptions$: Observable<QueryOptions>) {
```

Comment:
Should be private if that used external to this service.

Reply:
This part (data retrieval and `DataService`) is thoroughly modified and cleaner in #46 .